core/src/main/java/io/ultrabrew/metrics/data/ConcurrentMonoidIntTable.java

```diff
@@ -0,0 +1,469 @@
+
+package io.ultrabrew.metrics.data;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sun.misc.Unsafe;
+
+public abstract class ConcurrentMonoidIntTable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ConcurrentMonoidIntTable.class);
+
+  // Certainty required to meet the spec of probablePrime
+  private static final int DEFAULT_PRIME_CERTAINTY = 100;
+  private static final int DEFAULT_INITIAL_CAPACITY = 16;
+  private static final float DEFAULT_LOAD_FACTOR = 0.7f;
+  private static final int TAGSETS_MAX_INCREMENT = 131072; // 128k
+  private static final int RESERVED_FIELDS = 2;
+  protected static final long TABLE_MASK = 0x0FFFFFFF00000000L;
+  protected static final long SLOT_MASK = 0x00000000FFFFFFFFL;
+  protected static final long NOT_FOUND = 0x1000000000000000L;
+  protected static final Unsafe unsafe = UnsafeHelper.unsafe;
+  private static final long usedOffset;
+
+  ///CLOVER:OFF
+  // Turning off clover because Unsafe can't be safely mocked without crashing or otherwise
+  // hindering JVM, and Class can't be mocked
+  static {
+    try {
+      usedOffset = unsafe
+          .objectFieldOffset(ConcurrentMonoidIntTable.class.getDeclaredField("used"));
+    } catch (NoSuchFieldException e) {
+      throw new Error(e);
+    }
+  }
+  ///CLOVER:ON
+
+  /**
+   * The multiple int array objects to manipulate with the unsafe atomic operations.
+   */
+  protected List<int[]> tables;
+  /**
+   * Number of entries in the corresponding table in {@link #tables}
+   */
+  private List<AtomicInteger> recordCounts;
+  /**
+   * Max number of entries allowed in the corresponding table in {@link #tables}.
+   */
+  private final List<Integer> tableCapacities;
+  private final int recordSize;
+  private final int maxCapacity;
+  private final long[] identity;
+  private final int numAggFields;
+
+  private volatile int capacity;
+  protected volatile String[][] tagSets;
+  private volatile int used = 0;
+
+  /**
+   *  @param recordSize number of fields in a record
+   * @param maxCapacity maximum capacity of table in records. Table doesn't grow beyond this value.
+   * @param initialCapacity requested capacity of table in records
+   * @param identity monoid's identity for the agg fields
+   */
+  protected ConcurrentMonoidIntTable(final int recordSize, final int maxCapacity, int initialCapacity,
+      final long[] identity) {
+    this(identity.length, recordSize - identity.length, maxCapacity, initialCapacity, identity);
+  }
+
+  private ConcurrentMonoidIntTable(final int numAggFields, final int dataSize,
+      final int maxCapacity, int initialCapacity, final long[] identity) {
+
+    assert initialCapacity >= 0 : "Illegal initial capacity";
+    assert maxCapacity >= initialCapacity : "max capacity should be greater than the initial capacity";
```

Comment:
Please don't use `assert` in non-performance-critical code, almost no one ever runs JVM with assertions enabled.